File: countingValleys.js
function countingValleys(n, s) {
  let level = 0;
  let valleys = 0;

  for (let i = 0; i < n; i++) {
    if (level === 0 && s[i] === 'D') valleys += 1;

    if (s[i] === 'U') level += 1;
    else level -= 1;
  }

  return valleys;
}

console.log(countingValleys(8, 'UDDDUDUU'));
console.log(countingValleys(8, 'DDUUUUDD'));
console.log(countingValleys(12, 'DDUUDDUDUUUD'));

File: counting_valleys.py
def countingValleys(n, s):
    # initialize a variable level to zero
    level = 0
    # initialize a variable valleys to zero
    valleys = 0
    # loop through the steps in s
    for step in s:
        # if current step is an uphill one
        if step == 'U':
            # increment level by one
            level += 1
        # if current step is a downhill one
        else:
            # decrement level by one
            level -= 1
        # if level equals zero
        if level == 0:
            # increment valleys by one
            valleys += 1
    # return valleys
    return valleys
